File: app/src/main/java/ru/ra66it/updaterforspotify/data/repositories/SpotifyRepositoryImpl.java
package ru.ra66it.updaterforspotify.data.repositories;

import javax.inject.Inject;

import io.reactivex.Observable;
import ru.ra66it.updaterforspotify.data.network.SpotifyApi;
import ru.ra66it.updaterforspotify.domain.models.Spotify;
import ru.ra66it.updaterforspotify.domain.repositories.SpotifyRepository;


public class SpotifyRepositoryImpl implements SpotifyRepository {

    private SpotifyApi spotifyApi;

    @Inject
    public SpotifyRepositoryImpl(SpotifyApi spotifyApi) {
        this.spotifyApi = spotifyApi;
    }

    @Override
    public Observable<Spotify> getLatestSpotify() {
        return spotifyApi.getLatestSpotify();
    }
}

File: app/src/main/java/ru/ra66it/updaterforspotify/data/repositories/SpotifyRepositoryImpl.kt
package ru.ra66it.updaterforspotify.data.repositories

import ru.ra66it.updaterforspotify.data.network.SpotifyApi
import ru.ra66it.updaterforspotify.domain.model.Result
import ru.ra66it.updaterforspotify.domain.model.Spotify
import ru.ra66it.updaterforspotify.domain.repositories.SpotifyRepository
import ru.ra66it.updaterforspotify.presentation.utils.safeApiCall

class SpotifyRepositoryImpl (private val spotifyApi: SpotifyApi) : SpotifyRepository {

    override suspend fun getSpotify() = safeApiCall(
            call = { latestSpotify() }
    )

    private suspend fun latestSpotify(): Result<Spotify> {
        val response = spotifyApi.latestSpotifyAsync().await()
        return if (response.isSuccessful) {
            val body = checkNotNull(response.body())
            Result.Success(body)
        } else {
            val errorBody = checkNotNull(response.errorBody())
            Result.Error(Exception(errorBody.string()))
        }
    }

}
